### DISCUSSION TITLE
Migration strategy for a new mac

### DISCUSSION THREAD

**Discussion by mbierman** [Homebrew/discussions#1814] [Getting started]
Sorry for such a simple question... 

I am getting a new mac for the first time in a long time and I am wondering what the best way to migrate all the things I have installed with homebrew is.  I assume migration assistant won't bring it over?  So is there a way of getting a list of homebrew packages and, ideally creating a script or manifest so I can just mirror the same packages on the new machine?

**Comment by nandahkrishna** [comment]
You might want to check out `brew bundle`: https://github.com/Homebrew/homebrew-bundle. With `brew bundle dump` you can create a Brewfile with a list of all packages you installed, and you can use this to install packages on your new Mac.

**Comment by gromgit** [comment]
It's generally a good idea to do fresh installs of existing software on a new system, especially if (I assume) you're moving from Intel to Apple Silicon.

First back up your old Mac, if you haven't already done so. (This is just a precautionary measure, in case the next step goes pear-shaped for some bizarre reason.)

Then on your old Mac:
```
# Get the latest Homebrew software
brew update
# Dump your existing Homebrew install state
brew bundle dump --global --force
```
Now go over to your new Mac, and use Migration Assistant to pull over everything it can. After it's done, check to see if `$HOME/.Brewfile` exists; if not, copy it over.

Next, [install Homebrew afresh](https://brew.sh/) -- it should be writing to `/opt/homebrew` instead of `/usr/local`.

Finally, reinstall all your beloved formulae from fresh bottles:
```
brew bundle --global
```
You may have to wait a while, depending on how many formulae you'd previously installed.

**Reply by mbierman** [reply]
Thanks! Will most Brew packages work on M1 by now?

**Reply by gromgit** [reply]
Yes. Anything you use that isn't M1-ready should error out during the `brew bundle` install.

**Reply by mbierman** [reply]
Excellent. Thank you @gromgit and @nandahkrishna.

**Reply by mbierman** [reply]
@gromgit 

Is this correct? You said `/opt/homebrew` but it is going to `/usr/local` and I see no option to change it.

```
bigmac:~ michael$ /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
==> Checking for `sudo` access (which may request your password).
Password:
==> This script will install:
/usr/local/bin/brew
/usr/local/share/doc/homebrew
/usr/local/share/man/man1/brew.1
/usr/local/share/zsh/site-functions/_brew
/usr/local/etc/bash_completion.d/brew
/usr/local/Homebrew
```

is /opt/homebrew just on M1? Right now I'm reinstalling on my old machine.

**Reply by nandahkrishna** [reply]
`/opt/homebrew` is just for Apple Silicon. I assume your old machine has an Intel processor, so Homebrew will install into `/usr/local`.

**Reply by mbierman** [reply]
perfect, makes sense @nandahkrishna Thank you.

**Reply by mbierman** [reply]
One last question (I hope). My Brewfile looks like this (below).  Why is `brew "python@2", link: false` and `brew "git-ftp", link: false` ? Should I change that to reinstall?  I did a complete brew uninstall and reinstall and brew doctor says everything is good. Before I mess it up is there anything wrong with the Brewlist below?


```
tap "homebrew/bundle"
tap "homebrew/cask"
tap "homebrew/cask-drivers"
tap "homebrew/cask-fonts"
tap "homebrew/cask-versions"
tap "homebrew/core"
tap "ipinfo/tap"
tap "vitorgalvao/tiny-scripts"
brew "libssh2"
brew "curl"
brew "exiftags"
brew "fd"
brew "fswatch"
brew "gdbm"
brew "git-ftp", link: false
brew "grepip"
brew "iperf3"
brew "ipinfo-cli"
brew "jenv"
brew "jq"
brew "libxml2"
brew "python@2", link: false
brew "python@3.8"
brew "rsync"
brew "shellcheck"
brew "tag"
brew "tidy-html5"
brew "wakeonlan"
brew "wget"
brew "wireshark"
brew "vitorgalvao/tiny-scripts/cask-repair"
cask "aerial"
cask "blackhole-16ch"
cask "blackhole-2ch"
cask "keycastr"
```

**Reply by gromgit** [reply]
> Why is `brew "python@2", link: false` and `brew "git-ftp", link: false` ? Should I change that to reinstall?

Either you ran `brew unlink` on these two formulae, or Homebrew did the equivalent for some reason (can't think of a reason why though). You can manually `brew link` those formulae, then:
```
mv ~/.Brewfile ~/.Brewfile.old
brew bundle dump --global
```
then compare `Brewfile` and `Brewfile.old` to see if anything else changed. I suggest this because...

> is there anything wrong with the Brewlist below?

I don't see anything specifically "wrong" with your Brewfile, but assuming you haven't run `brew bundle --global` on your new Mac, you're guaranteed at least one error when you do: `python@2` no longer exists. If any of the other formulae and casks have been deprecated, disabled or removed entirely, you'll get errors for those too.

That said, `brew bundle --global` shouldn't hurt anything; you just may not get everything back that you used to have. To borrow the opening line from a long-running soap: "Like sands though the hourglass, so are the formulae of Homebrew."

**Reply by SMillerDev** [reply]
> brew "vitorgalvao/tiny-scripts/cask-repair"

This one is deprecated in favor of `brew bump-cask-version`. But I bet it won't break anything to install it.
